No std

perlindgren opened the pull request:
Support for building under `no_std`.

perlindgren committed:
no_std using core-error

perlindgren committed:
builds on no_std

perlindgren committed:
no_std support

helgoboss commented on src/newtype_macros.rs:
```diff
@@ -4,14 +4,16 @@
 pub struct TryFromGreaterError(pub(crate) ());
 
-impl std::error::Error for TryFromGreaterError {}
+impl core_error::Error for TryFromGreaterError {}
```
- [x] @helgoboss I'll test how this affects usage in my (std) crates.

helgoboss commented on src/newtype_macros.rs:
```diff
@@ -54,7 +56,9 @@ This function panics if `value` is greater than ", $max, "."
-                        format!("{} is not a valid value", value)
+                        // requires std
+                        //format!("{} is not a valid value", value)
+                        "Not a valid value"
```
@perlindgren Just like you I think that calling `new` with an invalid value is a usage fault and therefore justifies a panic. If anyone doesn't like the possibility of a panic and prefers a `Result`, then they could always use `.try_into()` which is also implemented for each newtype. Therefore I'm glad that the signature of `new` stays the same.

Concerning the removal of `format!`, maybe it would be a good idea to use it if `std` is available and not use it in `no_std` context. In `std` context, it's sometimes very helpful for error analysis to see the actual value that failed when getting a backtrace from production (it recently helped me to fix some user-reported bugs in [ReaLearn](https://github.com/helgoboss/realearn)). For the same reason I actually want to integrate the newtype name in the error message, but that's stuff for another PR. So what do you think about something like this (untested code)?

```rust
#[cfg(feature = "std")]
{
    assert!($name::is_valid(value), format!("{} is not a valid value", value));
}
#[cfg(feature = "no_std")]
{
    assert!($name::is_valid(value), "Not a valid value");
}
```

perlindgren replied:
Indeed you can have special case for `std`/`no_std`. In the tests I kept all the other dependencies to `std`, even though it might be possible to make also test working in `no_std` I don't think there is much benefit from it.

And as you mention, if there is a `try_into()` available then I have no issue with `new` panic`:ing.

I haven't yet tested the `helgoboss-midi` in a real application, I'll have a go at that later tonight integrating it in a crate `usbd-hid-midi` that I'm working on.

On another note, your use case is VST plugins and such. I'm also interested in that, have been playing around with JUCE a bit just to get a feel of the existing VST ecosystem. There are now some progress also on `baseview`/`baseplug` with `iced` based GUI bindings. I was looking into using `druid` but it turned out complicated. I have never coded for Reaper, but just saw that there is Rust bindings in the making. Very interesting! 

/Per

perlindgren committed:
newtype::new behind std/no_std variants

perlindgren commented:
I incorporated your suggestion in the PR. It passes the tests, and also compiles under `no_std`, so should be fine to merge.

helgoboss commented:
Great! It looks very good. Let me just try to migrate `reaper-rs` and ReaLearn to it and see how it goes.

helgoboss commented on Cargo.toml:
```diff
@@ -8,15 +8,17 @@ repository = "https://github.com/helgoboss/helgoboss-midi"
 edition = "2018"
-categories = ["data-structures", "multimedia"]
+categories = ["data-structures", "multimedia", "embedded", "no-std"]
 
```
```suggestion
[features]
default = ["std"]
std = []

```

helgoboss replied:
@perlindgren added the code manually

helgoboss commented on src/lib.rs:
```diff
@@ -1,3 +1,4 @@
+#![cfg_attr(not(test), no_std)]
```
```suggestion
#![cfg_attr(all(not(feature = "std"), not(test)), no_std)]
```

helgoboss reviewed (changes requested):
I tested it in my `std` context and provoked a panic by using an invalid value - but it didn't panic anymore. Reason: The `std` feature needs to be explicitly declared. The Rust API guidelines have a [section about this](https://rust-lang.github.io/api-guidelines/naming.html#feature-names-are-free-of-placeholder-words-c-feature).

I implemented the necessary changes and now it works. Added them as PR code suggestions. The only thing which I didn't know how to add (because of the triple backticks) was this additional `README.md` content. Could you add it?

    If you want to use it in a `no_std` context, add this instead:
    
    ```toml
    [dependencies]
    helgoboss-midi = { version = "0.1.0", default-features = false }
    ```

Would be cool if you could check once more if the `no_std` opt-in via `default-features = false` works that way. But I'm confident it will.

perlindgren commented:
> I tested it in my `std` context and provoked a panic by using an invalid value - but it didn't panic anymore. Reason: The `std` feature needs to be explicitly declared. The Rust API guidelines have a [section about this](https://rust-lang.github.io/api-guidelines/naming.html#feature-names-are-free-of-placeholder-words-c-feature).
> 
> I implemented the necessary changes and now it works. Added them as PR code suggestions. The only thing which I didn't know how to add (because of the triple backticks) was this additional `README.md` content. Could you add it?
> 
> ```
> If you want to use it in a `no_std` context, add this instead:
> 
> ```toml
> [dependencies]
> helgoboss-midi = { version = "0.1.0", default-features = false }
> ```
> ```
> 
> Would be cool if you could check once more if the `no_std` opt-in via `default-features = false` works that way. But I'm confident it will.

Hi Helgoboss.

I tried to connect to you over Matrix, but it indicates that the gitter bridge does not support private messaging. Can I reach you over Discord/Telegram or some other chat (I figure its easier to discuss directly).

perlindgren committed:
std as deafault feature

perlindgren commented:
So, at least it passes tests, compiles and links for the stm32f723 target I'm working with.

helgoboss commented:
@perlindgren Looks good, I'll merge this.

Oh, and on Discord you can reach me at helgoboss, `#1485`.

perlindgren commented:
> @perlindgren Looks good, I'll merge this.

Cool. I'm in the process of refactoring the `usdb-hid-midi` code, once that is done, I'll make some examples for using the `helgoboss-midi` abstractions. (The `usbd-hid-midi` is mostly about the descriptors and `usbd` interfacing.
